File: web/lib/modelo.ts
/**
 * Máscara por cultura sobre a saída do modelo de imagem.
 *
 * O PlantVillage empacota cultura e doença na mesma classe
 * (`Tomato___Early_blight`), então um modelo de 38 saídas responde *sempre*
 * uma das 38 - inclusive requeima de batata para uma foto de tomate. Mas o
 * agrônomo *sabe* o que plantou. Ao informar a cultura, as classes das outras
 * treze são zeradas e o que sobra é renormalizado.
 *
 * Isto existe antes do modelo de propósito: dá para escrever e testar toda a
 * interpretação da saída sem que exista um ONNX, e assim a fase 5 fica sendo
 * só "ligar o modelo na entrada". Nenhuma função aqui carrega, baixa ou
 * executa modelo nenhum - todas recebem um vetor pronto.
 *
 * AVISO PARA A FASE 5, sobre a recusa: a máscara DESTRÓI a confiança como
 * sinal de fora-da-distribuição. Renormalizar sobre as dez classes de tomate
 * faz qualquer imagem - inclusive uma mangueira - sair com confiança alta. O
 * sinal de recusa tem que ser calculado sobre os logits crus, antes de
 * `mascararPorCultura`; só a resposta usa a distribuição mascarada.
 */

import {
  CLASSES,
  IDS_FORA_DO_MODELO,
  POR_CULTURA,
  TOTAL_DE_CLASSES,
  type ClasseModelo,
} from "./contrato-modelo.ts";
import {
  CULTURAS,
  CULTURAS_FORA_DO_MODELO,
  SAUDAVEIS,
  type ClasseSaudavel,
  type Cultura,
} from "./base-conhecimento.ts";

export type Previsao = {
  classe: string;
  culturaId: string;
  /** Ficha a abrir, ou null quando o modelo respondeu "saudável". */
  doencaId: string | null;
  saudavel: boolean;
  /** Probabilidade JÁ renormalizada dentro da cultura. Não serve de sinal de
   *  fora-da-distribuição - ver o aviso no topo do arquivo. */
  confianca: number;
};

const SAUDAVEL_POR_CULTURA = new Map<string, ClasseSaudavel>(
  SAUDAVEIS.map((s) => [s.culturaId, s]),
);

const FORA_DO_MODELO = new Set<string>(IDS_FORA_DO_MODELO);

const MOTIVO_FORA_DO_MODELO = new Map<string, string>(
  CULTURAS_FORA_DO_MODELO.map((c) => [c.culturaId, c.motivo]),
);

/**
 * O modelo de imagem não tem saída nenhuma para esta cultura.
 *
 * Cana, café e algodão: o PlantVillage não as contém, nem doentes nem
 * saudáveis. Entraram na base porque são lavouras centrais no Brasil e o
 * escopo do app não é o do dataset, mas o modelo continua cego a elas.
 *
 * Isto NÃO é o mesmo que `indicesDaCultura(id).length === 0`, que também dá
 * zero para uma cultura que não existe - e essa seria um bug, não um fato a
 * comunicar. Quem for mostrar mensagem ao agrônomo tem que perguntar aqui.
 */
export function foraDoModelo(culturaId: string): boolean {
  return FORA_DO_MODELO.has(culturaId);
}

/**
 * Por que o modelo não cobre esta cultura, no texto curado da base.
 *
 * Vem do dado e não da interface: uma ausência explicada é decisão, uma
 * ausência silenciosa parece falha do aplicativo.
 */
export function motivoForaDoModelo(culturaId: string): string | null {
  return MOTIVO_FORA_DO_MODELO.get(culturaId) ?? null;
}

/**
 * O que o modelo pode responder para esta cultura.
 *
 * Vazio tanto para cultura inexistente quanto para cultura fora do modelo -
 * `foraDoModelo` é quem separa os dois casos.
 */
export function indicesDaCultura(culturaId: string): readonly number[] {
  return POR_CULTURA[culturaId] ?? [];
}

/**
 * A classe saudável da cultura, ou null quando o dataset não tem uma.
 *
 * `Orange` e `Squash` não têm: o PlantVillage só traz laranja com greening e
 * abóbora com oídio. Para essas duas o modelo nunca consegue responder "sem
 * doença", e quem chama isto não pode assumir que sempre vem alguma coisa.
 */
export function classeSaudavelDe(culturaId: string): ClasseSaudavel | null {
  return SAUDAVEL_POR_CULTURA.get(culturaId) ?? null;
}

/**
 * As doenças que a base cobre e o modelo não sabe reconhecer.
 *
 * Derivado de `classeModelo === null`, nunca escrito à mão: é o buraco do
 * modelo, e ele tem que se manter sozinho quando a base mudar. Para soja são
 * as duas principais do Brasil - o modelo só conhece soja saudável, então uma
 * folha com ferrugem asiática cai na classe saudável com confiança alta. É
 * exatamente isso que o laudo precisa dizer em voz alta.
 *
 * Para cana, café e algodão devolve a lista inteira de doenças da cultura,
 * porque o modelo não cobre nenhuma. Nesse caso a resposta útil não é esta
 * lista e sim `foraDoModelo`, que corta o fluxo antes da foto.
 */
export function doencasForaDoModelo(
  culturaId: string,
): { id: string; nome: string }[] {
  const cultura: Cultura | undefined = CULTURAS.find((c) => c.id === culturaId);
  return (cultura?.doencas ?? [])
    .filter((d) => d.classeModelo === null)
    .map((d) => ({ id: d.id, nome: d.nome }));
}

/**
 * Zera as classes que não são da cultura e renormaliza o que sobra.
 *
 * Devolve um vetor novo, do mesmo tamanho, cuja soma é 1 - a menos que a
 * cultura seja desconhecida ou que o modelo não tenha dado massa nenhuma às
 * classes dela, casos em que devolve tudo zero em vez de dividir por zero e
 * espalhar NaN pela tela.
 */
export function mascararPorCultura(
  probabilidades: ArrayLike<number>,
  culturaId: string,
): Float32Array {
  if (probabilidades.length !== TOTAL_DE_CLASSES) {
    throw new Error(
      `o modelo tem ${TOTAL_DE_CLASSES} saidas, recebi ${probabilidades.length}`,
    );
  }

  const mascarado = new Float32Array(TOTAL_DE_CLASSES);
  const indices = indicesDaCultura(culturaId);

  let total = 0;
  for (const i of indices) total += probabilidades[i];
  if (total <= 0) return mascarado;

  for (const i of indices) mascarado[i] = probabilidades[i] / total;
  return mascarado;
}

/**
 * A classe mais provável dentro da cultura informada.
 *
 * Empate resolve pelo menor índice, para dar sempre a mesma resposta ao mesmo
 * vetor. Devolve null quando a cultura é desconhecida ou quando o modelo não
 * deu massa nenhuma a ela - nesse caso não há resposta a dar, e inventar a
 * primeira classe da lista seria pior que admitir.
 */
export function prever(
  probabilidades: ArrayLike<number>,
  culturaId: string,
): Previsao | null {
  const mascarado = mascararPorCultura(probabilidades, culturaId);
  const indices = indicesDaCultura(culturaId);

  let melhor = -1;
  for (const i of indices) {
    if (melhor === -1 || mascarado[i] > mascarado[melhor]) melhor = i;
  }
  if (melhor === -1 || mascarado[melhor] <= 0) return null;

  const classe: ClasseModelo = CLASSES[melhor];
  return {
    classe: classe.classe,
    culturaId: classe.culturaId,
    doencaId: classe.doencaId,
    saudavel: classe.tipo === "saudavel",
    confianca: mascarado[melhor],
  };
}

File: app/modelo.py
"""Contrato entre a base de conhecimento e as 38 saidas do modelo de imagem.

O modelo devolve um vetor de 38 numeros. Sozinho ele nao diz nada: e preciso
saber que o indice 29 e `Tomato___Early_blight` e que essa classe abre a ficha
`tomate_pinta_preta`. Este modulo deriva esse mapeamento a partir da base
curada e o grava em data/contrato_modelo.json.

POR QUE ISTO EXISTE ANTES DO MODELO
-----------------------------------
Se a ordem usada no treino divergir desta em UMA posicao, a falha e silenciosa
e plausivel: uma foto de tomate vira `Tomato___Leaf_Mold` em vez de
`Late_blight`. Continua sendo doenca de tomate, a tela continua correta, e o
agronomo recebe o manejo errado. Nenhum teste de interface pega isso.

Entao a ordem e definida aqui, num lugar so, ANTES de qualquer treino - e o
notebook passa a conferir contra este arquivo em vez de decidir por conta.

A ORDEM
-------
`sorted()` sobre os rotulos, que e exatamente o que o
`torchvision.datasets.ImageFolder` produz ao varrer os diretorios do dataset.
E ordenacao por ponto de codigo, nao alfabetica, e isso tem consequencias que
parecem erro de digitacao mas nao sao:

  35 Tomato___Tomato_Yellow_Leaf_Curl_Virus
  36 Tomato___Tomato_mosaic_virus     <- depois, porque 'Y' (89) < 'm' (109)
  37 Tomato___healthy                 <- sempre no fim do bloco da cultura

Rodar:  python -m app.modelo
"""

import json

from app.db import RAIZ, carregar_json, validar

CAMINHO_CONTRATO = RAIZ / "data" / "contrato_modelo.json"

TOTAL_DE_CLASSES = 38

# Curado a mao: e decisao de engenharia, nao conteudo agronomico, e por isso
# nao mora em base_conhecimento.json. A fase 5 tem que reproduzir cada um
# destes valores em TypeScript - qualquer divergencia aqui vira acuracia
# perdida em silencio, porque o modelo recebe um tensor diferente do que viu
# no treino e mesmo assim responde com confianca.
PREPROCESSAMENTO = {
    "entrada": {
        "largura": 224,
        "altura": 224,
        "canais": 3,
        "layout": "NCHW",
        "tipo": "float32",
    },
    "redimensionamento": {
        # Transformacao de avaliacao padrao do torchvision: redimensiona o
        # lado menor para 256 mantendo proporcao, depois recorta 224 no centro.
        "lado_menor": 256,
        "recorte": "central",
        "metodo": "bilinear",
        "cantos_alinhados": False,
    },
    "normalizacao": {
        "escala": 255.0,
        "media": [0.485, 0.456, 0.406],
        "desvio": [0.229, 0.224, 0.225],
    },
    "saida": {
        # Logits, nunca softmax. O softmax fica do lado do TypeScript porque a
        # recusa da fase 5 precisa de temperatura e de energia, e as duas sao
        # impossiveis de calcular depois que o grafo ja normalizou.
        "tipo": "logits",
        "classes": TOTAL_DE_CLASSES,
    },
}

# Recusa: quando o app deve dizer "nao sei" em vez de responder.
#
# Os tres numeros sao NULOS ate a fase 4b: eles saem da curva risco-cobertura
# medida sobre imagens de validacao, e chutar um limiar seria pior que nao ter
# nenhum - daria ao agronomo uma recusa que nao significa nada. Enquanto forem
# nulos, o app calcula as pontuacoes e diz que o modelo nao esta calibrado.
RECUSA = {
    "temperatura": None,
    "limiar_msp": None,
    "limiar_energia": None,
    "calibrado_em": None,
    "medir_sobre": (
        "LOGITS CRUS, antes da mascara por cultura. A mascara renormaliza "
        "sobre as classes de uma cultura so, e isso faz QUALQUER imagem - "
        "inclusive uma mangueira apontada como tomate - sair com confianca "
        "alta. Calcular a recusa depois da mascara e o mesmo que nao ter "
        "recusa."
    ),
    "sabores_de_fora_da_distribuicao": [
        "cultura fora do dataset, DECLARADA na base - cana, cafe e algodao. "
        "O app nem chega a chamar o modelo quando o agronomo seleciona uma "
        "delas: `culturas_fora_do_modelo` resolve o caso antes da inferencia. "
        "Continuam sendo fora-da-distribuicao para o modelo, porque nada "
        "impede que uma folha de cafe seja fotografada com 'tomate' "
        "selecionado - e ai a recusa e a unica defesa.",
        "cultura fora do dataset e fora da base - feijao, trigo, manga. So a "
        "recusa pega, e depois dela a camada 3.",
        "cultura conhecida e classe desconhecida - doenca de soja, que o "
        "modelo so conhece saudavel. E o caso mais perigoso do app.",
    ],
}

# O que o notebook tem que garantir. Fica aqui, e nao so na prosa do notebook,
# para que a checagem seja copiavel e nao dependa de alguem lembrar.
CONTRATO_DE_TREINO = {
    "ordem_das_classes": (
        "sorted() sobre os rotulos, que e o que torchvision.datasets."
        "ImageFolder produz. Ordenacao por ponto de codigo, nao alfabetica."
    ),
    "assercao_obrigatoria": (
        "assert dataset.classes == [c['classe'] for c in contrato['classes']] "
        "- pega espelho do PlantVillage que 'conserta' o sublinhado final de "
        "Corn_(maize)___Common_rust_ ou troca por sublinhado o espaco de "
        "Tomato___Spider_mites Two-spotted_spider_mite."
    ),
    "metadata_onnx": (
        "gravar a lista de classes em metadata_props['classes'] como JSON. E "
        "a unica defesa contra service worker servindo um ONNX antigo junto de "
        "um bundle novo."
    ),
    "metricas": (
        "metricas.json tem que carregar classes_do_treino, para o repositorio "
        "poder afirmar que o modelo que voltou foi treinado nesta ordem."
    ),
}


def classes_canonicas(base: dict) -> list[str]:
    """As 38 classes na ordem em que o modelo as devolve.

    Uma linha so, de proposito: e a unica definicao de ordem do projeto
    inteiro. O TypeScript copia o resultado, nunca recalcula.
    """
    doencas = [d["classe_modelo"]
               for c in base["culturas"] for d in c["doencas"]
               if d["classe_modelo"] is not None]
    saudaveis = [s["classe_modelo"] for s in base["saudaveis"]]
    return sorted(doencas + saudaveis)


def montar() -> dict:
    base = carregar_json()
    validar(base)

    ordem = classes_canonicas(base)
    if len(ordem) != TOTAL_DE_CLASSES:
        raise ValueError(
            f"esperava {TOTAL_DE_CLASSES} classes, a base declara {len(ordem)}")

    # classe -> de onde ela veio
    dono: dict[str, dict] = {}
    for cultura in base["culturas"]:
        for d in cultura["doencas"]:
            if d["classe_modelo"] is not None:
                dono[d["classe_modelo"]] = {
                    "cultura_id": cultura["id"],
                    "tipo": "doenca",
                    "doenca_id": d["id"],
                }
    for s in base["saudaveis"]:
        dono[s["classe_modelo"]] = {
            "cultura_id": s["cultura_id"],
            "tipo": "saudavel",
            "doenca_id": None,
        }

    classes = [
        {"indice": i, "classe": classe, **dono[classe]}
        for i, classe in enumerate(ordem)
    ]

    # As culturas que o dataset nao contem de forma alguma. Ficam FORA de
    # `por_cultura`, e nao como lista vazia: uma lista vazia e indistinguivel
    # de cultura inexistente, e as duas situacoes pedem respostas diferentes
    # do app - "essa cultura nao existe" e um bug, "o modelo nao cobre cana"
    # e um fato permanente que o agronomo precisa ouvir antes de tirar a foto.
    fora_do_modelo = sorted(
        c["cultura_id"] for c in base["culturas_fora_do_modelo"])

    # A mascara por cultura, pre-calculada. O app zera tudo que nao esta na
    # lista da cultura escolhida e renormaliza - e o que impede o modelo de
    # responder requeima de batata para uma foto de tomate.
    por_cultura: dict[str, list[int]] = {
        c["id"]: [] for c in base["culturas"]
        if c["id"] not in set(fora_do_modelo)
    }
    for c in classes:
        por_cultura[c["cultura_id"]].append(c["indice"])

    return {
        "_comentario": (
            "GERADO AUTOMATICAMENTE por `python -m app.modelo`. Nao editar a "
            "mao. Liga cada saida do modelo de imagem a base de conhecimento: "
            "indice -> classe -> cultura e ficha. A ordem e sorted() sobre os "
            "rotulos, que e o que o ImageFolder do torchvision produz, e o "
            "notebook de treino tem que conferir contra ela antes de treinar."
        ),
        "gerado_por": "python -m app.modelo",
        "total_de_classes": TOTAL_DE_CLASSES,
        "classes": classes,
        "por_cultura": por_cultura,
        "culturas_fora_do_modelo": fora_do_modelo,
        "preprocessamento": PREPROCESSAMENTO,
        "recusa": RECUSA,
        "contrato_de_treino": CONTRATO_DE_TREINO,
        # Preenchidos quando o ONNX voltar do Colab. O sha256 e conferido no
        # download durante o build, para o app nunca servir um modelo que nao
        # seja o que foi medido no relatorio de validacao.
        "arquivo": {"nome": None, "sha256": None},
    }


def gerar() -> None:
    conteudo = montar()
    with open(CAMINHO_CONTRATO, "w", encoding="utf-8", newline="\n") as f:
        json.dump(conteudo, f, ensure_ascii=False, indent=2)
        f.write("\n")

    n_doenca = sum(1 for c in conteudo["classes"] if c["tipo"] == "doenca")
    n_saudavel = len(conteudo["classes"]) - n_doenca
    sem_saudavel = [
        c for c, indices in conteudo["por_cultura"].items()
        if not any(conteudo["classes"][i]["tipo"] == "saudavel" for i in indices)
    ]
    print(f"Contrato em {CAMINHO_CONTRATO}")
    print(f"  {len(conteudo['classes'])} classes: {n_doenca} de doenca, "
          f"{n_saudavel} saudaveis")
    print(f"  culturas sem classe saudavel: {', '.join(sem_saudavel)}")
    print(f"  culturas fora do modelo (sem saida nenhuma): "
          f"{', '.join(conteudo['culturas_fora_do_modelo'])}")


if __name__ == "__main__":
    gerar()
